### PR TITLE
AD400x: Add support for AD400x series

### PR DESCRIPTION
**Pull request by threexc**
- Add AD400x/Base.m class, refactor AD4020
- Add AD4021, AD4022
- Add AD4000, AD4004, AD4008
- Add AD4001, AD4005
- Add AD4002, AD4006, AD4010
- Add AD4003, AD4007, AD4011
- Update docs, metadata

### CODE REVIEW

**Comment by threexc** [comment]
Updated to match the current version of the ad400x driver.

File: +adi/+AD400x/Base.m [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                rValue = obj.getAttributeDouble('voltage0', 'offset', obj.isOutput);
+            else
+                rValue = NaN;
+            end
+        end
+    end
+
+    %% API Functions
+    methods (Hidden, Access = protected)
+        function setupInit(~)
```
Comment (ribdp) [inline +adi/+AD400x/Base.m:110]: Any attribute reads/writes performed before calling the system object won't be applied in the hardware. 

For the sampling rate attribute, should we cache the value in the property setter (if `ConnectedToDevice` is false), and invoke the setDeviceAttributeRAW method here in setupInit to apply it in the HW before the data capture?

**Comment by ribdp** [comment]
Just to confirm - have the drivers added as part of this PR been tested with the AD4020 example script (after changing the class name in the example script)?

File: +adi/+AD400x/Base.m (hunk 1 above)
Comment (threexc) [reply +adi/+AD400x/Base.m]: Hi, I've updated the commit to set the sampling frequency in this way in setupInit (based on what's happening in AD463x). Thanks for suggesting.

**Comment by threexc** [comment]
> Just to confirm - have the drivers added as part of this PR been tested with the AD4020 example script (after changing the class name in the example script)?

I've tested that example flow manually in the MATLAB terminal. The AD4020 example script doesn't quite work as-is right now, and I'm not sure why - it seems that when rx.Voltage() is called after rx(), it outputs this error:

`>> rx.Voltage()
Error using matlabshared.libiio.base/cstatus
Attribute read failed for : raw

Error in adi.common.Attribute/getAttributeRAW (line 145)
                cstatus(obj,status,['Attribute read failed for : ' attr]);

Error in adi.AD400x.Base/get.Voltage (line 83)
                rValue = obj.getAttributeRAW('voltage0', 'raw', obj.isOutput);
 `

I was planning on submitting a follow-up PR afterwards to remove that. Do you think that should be submitted first, or is there something I'm missing that explains why it won't work?

File: +adi/+AD400x/Base.m (hunk 1 above)
Comment (ribdp) [reply +adi/+AD400x/Base.m]: Thanks, the function looks good now

**Comment by threexc** [comment]
I suppose I could also just add that fix to this PR... :)

**Comment by ribdp** [comment]
> I suppose I could also just add that fix to this PR... :)

Oh yes, that'd work too. 
I was just going through the scripts trying to figure out why that error crops up in the first place...I wonder if this issue can be reproduced using a different IIO client (data capture, followed by attribute read). That would tell whether it originates from the MATLAB driver, or the firmware/Linux.

**Comment by threexc** [comment]
> > I suppose I could also just add that fix to this PR... :)
> 
> Oh yes, that'd work too. I was just going through the scripts trying to figure out why that error crops up in the first place...I wonder if this issue can be reproduced using a different IIO client (data capture, followed by attribute read). That would tell whether it originates from the MATLAB driver, or the firmware/Linux.

I think it's the kernel driver - if I do rx() in MATLAB and then try to read the raw voltage0 value, I get `ERROR: Device or resource busy (16)`. But this makes sense, since a buffer is enabled for rx() and you can't do the single conversion read when that's the case. 

I've pushed an update that removes the line from the example script. I'm wondering if the Voltage() function should really be in there at all, though.

**Comment by ribdp** [comment]
> > > I suppose I could also just add that fix to this PR... :)
> > 
> > 
> > Oh yes, that'd work too. I was just going through the scripts trying to figure out why that error crops up in the first place...I wonder if this issue can be reproduced using a different IIO client (data capture, followed by attribute read). That would tell whether it originates from the MATLAB driver, or the firmware/Linux.
> 
> I think it's the kernel driver - if I do rx() in MATLAB and then try to read the raw voltage0 value with iio_info, I get `ERROR: Device or resource busy (16)`. But this makes sense, since a buffer is enabled for rx() and you can't do the single conversion read when that's the case.
> 
> I've pushed an update that removes the line from the example script. I'm wondering if the Voltage() function should really be in there at all, though.

I think it should be possible to destroy the buffer for rx channels, and then do attribute reads/writes. Something like this, perhaps - https://github.com/analogdevicesinc/ToolboxCommon/blob/be769e5517eea786eb4c0b32e231d6a03be6c105/RxTx.m#L195 ?

**Comment by threexc** [comment]
> > > > I suppose I could also just add that fix to this PR... :)
> > > 
> > > 
> > > Oh yes, that'd work too. I was just going through the scripts trying to figure out why that error crops up in the first place...I wonder if this issue can be reproduced using a different IIO client (data capture, followed by attribute read). That would tell whether it originates from the MATLAB driver, or the firmware/Linux.
> > 
> > 
> > I think it's the kernel driver - if I do rx() in MATLAB and then try to read the raw voltage0 value with iio_info, I get `ERROR: Device or resource busy (16)`. But this makes sense, since a buffer is enabled for rx() and you can't do the single conversion read when that's the case.
> > I've pushed an update that removes the line from the example script. I'm wondering if the Voltage() function should really be in there at all, though.
> 
> I think it should be possible to destroy the buffer for rx channels, and then do attribute reads/writes. Something like this, perhaps - https://github.com/analogdevicesinc/ToolboxCommon/blob/be769e5517eea786eb4c0b32e231d6a03be6c105/RxTx.m#L195 ?

I've made a local change to inherit the RxTx contents and tried to call releaseChanBuffers/destroyBuf, but both give "Cannot access method" results (which seems reasonable). I'm not overly familiar with MATLAB's object-oriented quirks though - do you know if I'm missing something with the use of those functions, or do you mean that I should re-define them inside the AD400x Base.m module?

**Comment by ribdp** [comment]
> > > > > I suppose I could also just add that fix to this PR... :)
> > > > 
> > > > 
> > > > Oh yes, that'd work too. I was just going through the scripts trying to figure out why that error crops up in the first place...I wonder if this issue can be reproduced using a different IIO client (data capture, followed by attribute read). That would tell whether it originates from the MATLAB driver, or the firmware/Linux.
> > > 
> > > 
> > > I think it's the kernel driver - if I do rx() in MATLAB and then try to read the raw voltage0 value with iio_info, I get `ERROR: Device or resource busy (16)`. But this makes sense, since a buffer is enabled for rx() and you can't do the single conversion read when that's the case.
> > > I've pushed an update that removes the line from the example script. I'm wondering if the Voltage() function should really be in there at all, though.
> > 
> > 
> > I think it should be possible to destroy the buffer for rx channels, and then do attribute reads/writes. Something like this, perhaps - https://github.com/analogdevicesinc/ToolboxCommon/blob/be769e5517eea786eb4c0b32e231d6a03be6c105/RxTx.m#L195 ?
> 
> I've made a local change to inherit the RxTx contents and tried to call releaseChanBuffers/destroyBuf, but both give "Cannot access method" results (which seems reasonable). I'm not overly familiar with MATLAB's object-oriented quirks though - do you know if I'm missing something with the use of those functions, or do you mean that I should re-define them inside the AD400x Base.m module?

On second thoughts - I'd say we can leave it be, and drop the `rx.Voltage()` for now. There's probably a way (can't think of one off the top of my head atm) to make the single conversion read work in that manner - but it's likely to be kinda messy, breaking the software layering in a dozen places. 

It's better if I create an issue, and think through any refactoring that needs to be done to have an elegant solution that works for all the supported parts.

**Comment by threexc** [comment]
I pushed a new version that removes the Voltage() function entirely, but left VoltageScale() and VoltageOffset(), which both work OK.

**Comment by threexc** [comment]
Should be good to go now. The version I pushed yesterday fixes datatypes for the AD4001 and AD4005, which I had mistakenly listed as int16 when refactoring to use the AD400x base class.

**Comment by ribdp** [comment]
Lgtm

**Review by ribdp — APPROVED** [review]

**Comment by threexc** [comment]
Thanks!

**Pull request merged**